src/lib.rs

```diff
@@ -1 +1,452 @@
+        match original::Constants::from_elements(&elements.into()) {
+            Ok(state) => Ok(Handle::new(Constants::new_wrap(state))),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn from_elements_afspc_compatibility_mode(
+        elements: ElementS,
+    ) -> Result<Handle<Constants>, SgpError> {
+        match original::Constants::from_elements_afspc_compatibility_mode(&elements.into()) {
+            Ok(state) => Ok(Handle::new(Constants::new_wrap(state))),
+            Err(err) => Err(err.into()),
+        }
+    }
+    fn propagate(&self, t: f64) -> Result<Prediction, SgpError> {
+        match self.0.propagate(t) {
+            Ok(prediction) => Ok(prediction.into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn propagate_afspc_compatibility_mode(&self, t: f64) -> Result<Prediction, SgpError> {
+        match self.0.propagate_afspc_compatibility_mode(t) {
+            Ok(prediction) => Ok(prediction.into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn initial_state(&self) -> Option<Handle<ResonanceState>> {
+        match self.0.initial_state() {
+            Some(rs) => Some(Handle::new(ResonanceState::new(rs))),
+            None => None,
+        }
+    }
+
+    fn propagate_from_state(
+        &self,
+        t: f64,
+        state: Option<Handle<ResonanceState>>,
+        afspc_compatibility_mode: bool,
+    ) -> Result<Prediction, SgpError> {
+        let mut rs = *state
+            .expect("Unable to get resonance state")
+            .0
+            .lock()
+            .expect("err");
+
+        match self
+            .0
+            .propagate_from_state(t, Some(&mut rs), afspc_compatibility_mode)
+        {
+            Ok(prediction) => Ok(prediction.into()),
+            Err(err) => Err(err.into()),
+        }
+    }
+}
+
+struct Elements(original::Elements);
+
+impl Elements {
+    fn new(state: original::Elements) -> Self {
+        Elements(state)
+    }
+}
+
+impl sgp4::Elements for Elements {
+    fn from_tle(
+        object_name: Option<String>,
+        line1: String,
+        line2: String,
+    ) -> Result<Handle<Elements>, SgpError> {
+        match original::Elements::from_tle(object_name, line1.as_bytes(), line2.as_bytes()) {
+            Ok(elements) => Ok(Handle::new(Elements::new(elements))),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn epoch(&self) -> f64 {
+        self.0.epoch()
+    }
+
+    fn epoch_afspc_compatibility_mode(&self) -> f64 {
+        self.0.epoch_afspc_compatibility_mode()
+    }
+}
+
+impl From<original::Prediction> for Prediction {
+    fn from(prediction: original::Prediction) -> Self {
+        let original::Prediction { position, velocity } = prediction;
+        Prediction {
+            position: (position[0], position[1], position[2]),
+            velocity: (velocity[0], velocity[1], velocity[2]),
+        }
```

Comment:
If you want, you should be able to use destructuring here.

```suggestion
        let original::Prediction { position: [p1, p2, p3], velocity: [v1, v2, v3] } = prediction;
        Prediction {
            position: (p1, p2, p3),
            velocity: (v1, v2, v3),
        }
```

The benefit is that you know `position` and `velocity` are `[f64; 3]` at compile time, so we can avoid the (theoretically fallible) indexing operation. It may or may not also be easier to read after `rustfmt` has done its thing.